// eslint-plugin-sonarjs/cjs/S5973/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S5973/javascript
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const index_js_1 = require("../helpers/index.js");
const meta = __importStar(require("./generated-meta.js"));
const dependencies_js_1 = require("../helpers/package-jsons/dependencies.js");
exports.rule = {
    meta: (0, index_js_1.generateMeta)(meta, {
        messages: {
            stable: 'Make your tests stable so that they pass on the first try, or remove the flaky ones.',
        },
    }),
    create(context) {
        const describes = [];
        const hasJest = hasJestDependency(context);
        return {
            /**
             * We use the stack approach to check for Mocha retries inside describe blocks,
             * and implicitly test cases.
             */
            CallExpression(node) {
                if (hasJestRetry(context, node, hasJest)) {
                    report(context, node);
                    return;
                }
                if (index_js_1.Mocha.isDescribeCase(node)) {
                    describes.push(node);
                    return;
                }
                if (describes.length > 0) {
                    checkMochaRetries(context, node);
                }
            },
            'CallExpression:exit': (node) => {
                if (index_js_1.Mocha.isDescribeCase(node)) {
                    describes.pop();
                }
            },
            'Program:exit': () => {
                describes.length = 0;
            },
        };
    },
};
function hasJestRetry(context, node, hasJest) {
    const callExpressionName = (0, index_js_1.getFullyQualifiedName)(context, node);
    return (callExpressionName === 'jest.retryTimes' ||
        (hasJest && (0, index_js_1.isMethodInvocation)(node, 'jest', 'retryTimes', 1)));
}
function hasJestDependency(context) {
    const dependencies = (0, dependencies_js_1.getDependenciesSanitizePaths)(context);
    return dependencies.has('jest');
}
/**
 * Flag if `this.retries()`
 */
function checkMochaRetries(context, node) {
    const callee = node.callee;
    if (callee.type === 'MemberExpression' &&
        callee.object.type === 'ThisExpression' &&
        (0, index_js_1.isIdentifier)(callee.property, 'retries')) {
        report(context, node);
    }
}
function report(context, node) {
    context.report({
        messageId: 'stable',
        node,
    });
}
